components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOArtifactResolver.java

```diff
@@ -44,53 +60,175 @@
-                    // Build Response.
-                    ResponseBuilder respBuilder = SAMLSSOUtil.getResponseBuilder();
-                    if (respBuilder != null) {
-                        response = respBuilder.buildResponse(artifactInfo.getAuthnReqDTO(),
-                                artifactInfo.getSessionID(), artifactInfo.getInitTimestamp(),
-                                artifactInfo.getAssertionID());
-                    } else {
-                        throw new ArtifactBindingException("Could not create a ResponseBuilder for SAML artifact " +
-                                "resolution.");
-                    }
+            SAML2ArtifactInfo artifactInfo = saml2ArtifactInfoDAO.getSAMLArtifactInfo(sourceIDString,
+                    messageHandlerString);
+
+            if (artifactInfo != null && validateArtifactResolve(artifactResolve, artifactInfo)) {
+                // Building Response.
+                ResponseBuilder respBuilder = SAMLSSOUtil.getResponseBuilder();
+                if (respBuilder != null) {
+                    response = respBuilder.buildResponse(artifactInfo.getAuthnReqDTO(), artifactInfo.getSessionID(),
+                            artifactInfo.getInitTimestamp(), artifactInfo.getAssertionID());
+
                 } else {
-                    log.warn("Artifact validity period (" + artifactInfo.getExpTimestamp() + ") has been " +
-                            "exceeded for artifact: " + artifact);
+                    throw new ArtifactBindingException("Could not create a ResponseBuilder for SAML2 artifact " +
+                            "resolution.");
                 }
             }
+
+            artifactResponse = buildArtifactResponse(response, artifactResolve, artifactInfo.getAuthnReqDTO());
+
         } catch (IdentityException e) {
             throw new ArtifactBindingException("Error while building response for SAML2 artifact: " + artifact, e);
         }
         catch (Base64DecodingException e) {
             throw new ArtifactBindingException("Error while Base64 decoding SAML2 artifact: " + artifact, e);
         }
 
-        return response;
+        return artifactResponse;
+    }
+
+    /**
+     * Validate the artifact resolve request for issuer, validity period and signature.
+     *
+     * @param artifactResolve ArtifactResolve object to be validated.
+     * @throws IdentityException
+     * @throws ArtifactBindingException
+     */
+    private boolean validateArtifactResolve(ArtifactResolve artifactResolve, SAML2ArtifactInfo artifactInfo)
+            throws IdentityException, ArtifactBindingException {
+
+        boolean isValid = true;
+
+        // Checking for artifactResolve validity period.
+        DateTime currentTime = new DateTime();
+        if (!artifactInfo.getExpTimestamp().isAfter(currentTime)) {
+            log.warn("Artifact validity period (" + artifactInfo.getExpTimestamp() + ") has been " +
```

Comment:
Will this warn log be printed too often ?

Reply:
This will be printed if the Artifact validity period is not correct in resolution request. This will be helpful to identify the reason for a particular app to fail. On the other hand, someone could use this to clog the logs. Maybe we could make it a debug log. WDYT?